From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Thu, 4 May 2017 15:58:57 -0400
Subject: [PATCH 061/100] slub: Add support for verifying slab sanitization

This is an extension to the sanitization feature in PaX for when
sacricifing more performance for security is acceptable.

The initial version from Daniel Micay was relying on PAGE_SANITIZE. It
now relies on upstream's init_on_free.

[nicolas.bouchinet@ssi.gouv.fr: Should not conflict with commit 520a688a2edfddba9]
[levente@leventepolyak.net: Adapt to init bool move into slab_post_alloc_hook() in f09b59ae4414b0dc]

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
Signed-off-by: Thibaut Sautereau <thibaut.sautereau@ssi.gouv.fr>
Signed-off-by: Nicolas Bouchinet <nicolas.bouchinet@ssi.gouv.fr>
Signed-off-by: Levente Polyak <levente@leventepolyak.net>
---
 mm/slub.c                  | 38 +++++++++++++++++++++++++++++++++++---
 security/Kconfig.hardening |  8 ++++++++
 2 files changed, 43 insertions(+), 3 deletions(-)

diff --git a/mm/slub.c b/mm/slub.c
index 38df8ebb43bdb..33542e9bad3d5 100644
--- a/mm/slub.c
+++ b/mm/slub.c
@@ -245,6 +245,12 @@ static inline bool kmem_cache_debug(struct kmem_cache *s)
 	return kmem_cache_debug_flags(s, SLAB_DEBUG_FLAGS);
 }
 
+static inline bool has_sanitize_verify(struct kmem_cache *s)
+{
+	return IS_ENABLED(CONFIG_SLAB_SANITIZE_VERIFY) &&
+	       slab_want_init_on_free(s);
+}
+
 void *fixup_red_left(struct kmem_cache *s, void *p)
 {
 	if (kmem_cache_debug_flags(s, SLAB_RED_ZONE))
@@ -2678,7 +2684,7 @@ bool slab_free_hook(struct kmem_cache *s, void *x, bool init,
 		 */
 		set_orig_size(s, x, orig_size);
 
-		if (s->ctor)
+		if (!IS_ENABLED(CONFIG_SLAB_SANITIZE_VERIFY) && s->ctor)
 			s->ctor(x);
 	}
 	/* KASAN might put x into memory quarantine, delaying its reuse. */
@@ -2749,7 +2755,7 @@ static inline void *setup_object(struct kmem_cache *s, void *object)
 {
 	setup_object_debug(s, object);
 	object = kasan_init_slab_obj(s, object);
-	if (unlikely(s->ctor)) {
+	if (unlikely(s->ctor) && !has_sanitize_verify(s)) {
 		kasan_unpoison_new_object(s, object);
 		s->ctor(object);
 		kasan_poison_new_object(s, object);
@@ -4570,7 +4576,7 @@ static __fastpath_inline
 bool slab_post_alloc_hook(struct kmem_cache *s, gfp_t flags, size_t size,
 			  void **p, const struct slab_alloc_context *ac)
 {
-	bool init = slab_want_init_on_alloc(flags, s);
+	bool init = slab_want_init_on_alloc(flags, s) && !has_sanitize_verify(s);
 	unsigned int zero_size = s->object_size;
 	gfp_t init_flags = flags & gfp_allowed_mask;
 	bool kasan_init = false;
@@ -4933,6 +4939,17 @@ static __fastpath_inline void *slab_alloc_node(struct kmem_cache *s,
 
 	maybe_wipe_obj_freeptr(s, object);
 
+	if (has_sanitize_verify(s) && object) {
+		/* KASAN hasn't unpoisoned the object yet (this is done in the
+		 * post-alloc hook), so let's do it temporarily.
+		 */
+		kasan_unpoison_new_object(s, object);
+		BUG_ON(memchr_inv(object, 0, s->object_size));
+		if (s->ctor)
+			s->ctor(object);
+		kasan_poison_new_object(s, object);
+	}
+
 out:
 	/*
 	 * In case this fails due to memcg_slab_post_alloc_hook(),
@@ -7387,6 +7404,21 @@ static bool __kmem_cache_alloc_bulk(struct kmem_cache *s, gfp_t flags,
 		stat_add(s, ALLOC_SLOWPATH, i);
 	}
 
+	if (has_sanitize_verify(s)) {
+		int j;
+
+		for (j = 0; j < i; j++) {
+			/* KASAN hasn't unpoisoned the object yet (this is done in the
+			 * post-alloc hook), so let's do it temporarily.
+			 */
+			kasan_unpoison_new_object(s, p[j]);
+			BUG_ON(memchr_inv(p[j], 0, s->object_size));
+			if (s->ctor)
+				s->ctor(p[j]);
+			kasan_poison_new_object(s, p[j]);
+		}
+	}
+
 	return true;
 
 error:
diff --git a/security/Kconfig.hardening b/security/Kconfig.hardening
index a1a14b208b859..ed5cfe8e744e7 100644
--- a/security/Kconfig.hardening
+++ b/security/Kconfig.hardening
@@ -216,6 +216,14 @@ config PAGE_SANITIZE_VERIFY
 	  When init_on_free is enabled, verify that newly allocated pages
 	  are zeroed to detect write-after-free bugs.
 
+config SLAB_SANITIZE_VERIFY
+	bool "Verify sanitized SLAB allocations"
+	default y
+	depends on !KASAN
+	help
+	  When init_on_free is enabled, verify that newly allocated slab
+	  objects are zeroed to detect write-after-free bugs.
+
 endmenu
 
 menu "Bounds checking"

